adapters/insticator/insticator.go

```diff
@@ -0,0 +1,526 @@
+		}
+	}
+
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		videoCopy, err := validateVideoParams(imp.Video)
+
+		imp.Video = videoCopy
+
+		if err != nil {
+			return openrtb2.Imp{}, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+	return imp, nil
+}
+
+func validateVideoParams(video *openrtb2.Video) (*openrtb2.Video, error) {
+	videoCopy := *video
+	if (videoCopy.W == nil || *videoCopy.W == 0) ||
+		(videoCopy.H == nil || *videoCopy.H == 0) ||
+		videoCopy.MIMEs == nil {
+
+		return nil, &errortypes.BadInput{
+			Message: "One or more invalid or missing video field(s) w, h, mimes",
+		}
+	}
+
+	// Validate optional parameters and remove invalid ones
+	cleanedVideo, err := validateOptionalVideoParams(&videoCopy)
+	if err != nil {
+		return nil, err
+	}
+
+	return cleanedVideo, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]InsticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, CALLER)
+
+	return json.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return roundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+// roundTo4Decimals function
+func roundTo4Decimals(amount float64) float64 {
+	return math.Round(amount*10000) / 10000
+}
+
+func validateOptionalVideoParams(video *openrtb2.Video) (*openrtb2.Video, error) {
+	v := reflect.ValueOf(video).Elem()
+	t := v.Type()
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		fieldValue := v.Field(i)
+
+		// Convert the field name to camelCase for matching in optionalVideoParams map
+		jsonTag := field.Tag.Get("json")
+		if jsonTag == "" {
+			jsonTag = toCamelCase(field.Name)
+		}
+
+		// Skip fields that are not in optionalVideoParams
+		validator, exists := optionalVideoParams[jsonTag]
+		if !exists {
+			continue
+		}
+
+		// Check if the field value is zero/nil and skip if true
+		if isZeroOrNil(fieldValue) {
+			continue
+		}
+
+		// Validate the field value
+		if !validator(fieldValue.Interface()) {
+			// If invalid, set the field to zero value
+			fieldValue.Set(reflect.Zero(fieldValue.Type()))
+		}
+	}
+	return video, nil
+}
+
+// Helper function to convert field name to camelCase
+func toCamelCase(s string) string {
+	if s == "" {
+		return s
+	}
+	return strings.ToLower(string(s[0])) + s[1:]
+}
+
+// Helper function to check if a value is zero or nil
+func isZeroOrNil(value reflect.Value) bool {
+	switch value.Kind() {
```

Comment:
based on https://github.com/prebid/prebid-server/pull/3806/files#r1705364786 comment this function won't be needed